tests/dialects/test_dlti.py

```diff
@@ -0,0 +1,264 @@
+    "entries",
+    [
+        [("key1", "value1"), ("key2", 23), (i32, FloatAttr(9.3, Float32Type()))],
+        [("k", IntAttr(23)), (i32, FloatAttr(2.4, Float32Type()))],
+    ],
+)
+def test_target_device_spec(
+    entries: list[tuple[StringAttr | TypeAttribute | str, Attribute | str | int]],
+):
+    # Test initialisation from a dictionary
+    generic_specification_test(
+        TargetDeviceSpecAttr, {e[0]: e[1] for e in entries}, entries
+    )
+    # Test initialisation from an array attribute of data layout entries
+    generic_specification_test(
+        TargetDeviceSpecAttr,
+        ArrayAttr([DataLayoutEntryAttr(e[0], e[1]) for e in entries]),
+        entries,
+    )
+
+
+@pytest.mark.parametrize(
+    "entries",
+    [
+        [("key1", "value1"), ("key2", 23), (i32, "test")],
+        [("k", IntAttr(23)), (i32, FloatAttr(2.4, Float32Type()))],
+    ],
+)
+def test_target_system_spec(
+    entries: list[tuple[StringAttr | TypeAttribute | str, Attribute | str | int]],
+):
+    # Test initialisation from a dictionary
+    generic_specification_test(
+        TargetSystemSpecAttr, {e[0]: e[1] for e in entries}, entries
+    )
+    # Test initialisation from an array attribute of data layout entries
+    generic_specification_test(
+        TargetSystemSpecAttr,
+        ArrayAttr([DataLayoutEntryAttr(e[0], e[1]) for e in entries]),
+        entries,
+    )
+
+
+@pytest.mark.parametrize(
+    "entries",
+    [
+        [("key1", "value1"), ("key2", 23), (i32, 17)],
+        [("k", IntAttr(23)), (i32, FloatAttr(2.4, Float32Type()))],
+    ],
+)
+def test_map_attr(
+    entries: list[tuple[StringAttr | TypeAttribute | str, Attribute | str | int]],
+):
+    # Test initialisation from a dictionary
+    generic_specification_test(MapAttr, {e[0]: e[1] for e in entries}, entries)
+    # Test initialisation from an array attribute of data layout entries
+    generic_specification_test(
+        MapAttr, ArrayAttr([DataLayoutEntryAttr(e[0], e[1]) for e in entries]), entries
+    )
+
+
+def test_map_attr_embedded_dict():
+    m = MapAttr({"key": {"key": {"key": "value"}}})
+    m.verify()
+    assert isinstance(m.entries.data[0].value, MapAttr)
+    assert isinstance(m.entries.data[0].value.entries.data[0].value, MapAttr)
+    embedded_contents = m.entries.data[0].value.entries.data[0].value
+    assert isinstance(embedded_contents.entries.data[0].key, StringAttr)
+    assert isinstance(embedded_contents.entries.data[0].value, StringAttr)
+
+
+def test_duplicate_data_layout_spec_entries():
+    with pytest.raises(VerifyException):
+        entry = DataLayoutSpecAttr(
+            ArrayAttr([DataLayoutEntryAttr("k", "v"), DataLayoutEntryAttr("k", 12)])
+        )
+        entry.verify()
+
+
+def test_duplicate_target_device_spec_entries():
+    with pytest.raises(VerifyException):
+        entry = TargetDeviceSpecAttr(
+            ArrayAttr([DataLayoutEntryAttr("k", "v"), DataLayoutEntryAttr("k", 12)])
+        )
+        entry.verify()
```

Comment:
these should not be necessary, as attributes are automatically verified in the init
```suggestion
```